<!-- {summary} --><a name="summary"></a>
<h2>Summary</h2>
<p>The Wellcome Trust Sanger Institute (WTSI) assembly.</p>
<!-- {summary} -->

<!-- {assembly} --><a name="assembly"></a>
<h2>Assembly</h2>
<p>The <em>H.&nbsp;bakeri</em> genome was sequenced by the <a href="http://www.nematodes.org/">Blaxter  laboratory at University of Edinburgh</a>. The data represented here is version nHp.1.0, obtained from <a href="http://nematodes.org/genomes/heligmosomoides_polygyrus/">nematodes.org</a>.</p>
<!-- {assembly} -->

<!-- {annotation} --><a name="annotation"></a>
<h2>Annotation</h2>
<p>The genome was annotated by the <a href="http://www.sanger.ac.uk/research/projects/parasitegenomics/">Parasite Genomics group at the Wellcome Trust Sanger Institute</a> and WormBase ParaSite, as part of the 50 Helminth Genomes Initiative (Helminth Genomes Consortium, unpublished). In in-house pipeline was developed that used <a href="http://europepmc.org/abstract/MED/22192575">MAKER</a> to generate high-quality annotations by integrating evidence from multiple sources: ab initio gene predictions from <a href="http://europepmc.org/abstract/MED/16845043">AUGUSTUS</a>, <a href="http://europepmc.org/abstract/MED/18757608">GeneMark-ES</a>, and <a href="http://europepmc.org/abstract/MED/15144565">SNAP</a>; projected annotation from <em>C.&nbsp;elegans</em> (using <a href="http://europepmc.org/abstract/MED/21653517">GenBlastG</a>) and the taxonomically nearest reference helminth genome (using <a href="http://europepmc.org/abstract/MED/21306991">RATT</a>); and ESTs, mRNAs and proteins from related organisms aligned to the genome using BLAST, with refinement of alignments using <a href="http://europepmc.org/abstract/MED/15713233">Exonerate</a>.</p>
<!-- {annotation} -->

