<!-- {summary} --><a name="summary"></a>
<h2>Summary</h2>
<p>The Wellcome Trust Sanger Institute (WTSI) assembly.</p>
<!-- {summary} -->

<!-- {assembly} --><a name="assembly"></a>
<h2>Assembly</h2>
<p>The genome was produced by the  <a href="http://www.sanger.ac.uk/research/projects/parasitegenomics/">Parasite Genomic group at the Wellcome Trust Sanger Institute</a>, using Illumina paired-end sequencing followed by an in-house genome assembly pipeline comprising various steps, including contig assembly, scaffolding,  gap-filling and error-correction (WTSI Strongyloididae Nematode Genomes Project, unpublished).</p>
<!-- {assembly} -->

<!-- {annotation} --><a name="annotation"></a>
<h2>Annotation</h2>
<p>The genome was annotated by the <a href="http://www.sanger.ac.uk/research/projects/parasitegenomics/">Parasite Genomics group at the Wellcome Trust Sanger Institute</a> as part of the Strongyloididae Nematode Genomes Project (unpublished). In in-house automatic pipeline (based around <a href="http://europepmc.org/abstract/MED/22192575">MAKER</a>) was used to generate first-pass gene models, and this was followed by a round of targeted manual curation of selected gene families. </p>
<!-- {annotation} -->

