
      <h3>Summary</h3>
    
      <table class="ss tint species-stats">
        <tr class="bg2">
          <td class="data">Assembly:</td>
          <td class="value">T_regenti_v1_0_4,  <a href="http://www.ebi.ac.uk/ena/data/view/GCA_000950905.1">GCA_000950905.1</a></td>
        </tr>
        <tr>
          <td class="data">Database version:</td>
          <td class="value">84.1</td>
        </tr>
      <tr class="bg2">
          <td class="data">Base Pairs:</td>
          <td class="value">701,762,036</td>
      </tr><tr>
          <td class="data">Genome Size:</td>
          <td class="value">701,762,036</td>
      </tr>
    <tr class="bg2">
          <td class="data">Data source:</td>
          <td class="value"><a href="http://www.sanger.ac.uk">Wellcome Trust Sanger Institute</a><br /></td>
      </tr>
      <tr> <td class="data">Genebuild version:</td> <td class="value">2014-06-WormBase</td> </tr></table>
          <h3>Gene counts</h3>
          <table class="ss tint species-stats">
        <tr class="bg2">
              <td class="data"><span class="glossary_mouseover">Coding genes<span class="floating_popup"><p>Genes and/or transcript that contains an open reading frame (ORF).</p></span></span>:</td>
              <td class="value">22,185</td>
              </tr>
            <tr>
              <td class="data"> <span class="glossary_mouseover">Gene transcripts<span class="floating_popup">Nucleotide sequence resulting from the transcription of the genomic DNA to mRNA. One gene can have different transcripts or splice variants resulting from the alternative splicing of different exons in genes.</span></span>:</td>
              <td class="value">22,185</td>
              </tr>
            
          </table>
        
          </table>
        <h3>Coordinate Systems</h3>
<table class="ss tint species-stats">
            <tr class="bg2"> 
            <td class="data">scaffold</td>
            <td class="value">188369 sequences</td>
            </tr></table>
